### PR TITLE
Inverted Mode

### PR DESCRIPTION
**Pull request by cassidoxa**
Notes:

- Added inverted ROM changes or applied ER ROM fixes where appropriate
    -Includes text changes for intro and spawn points
- Added new module InvertedRegions.py which reflects inverted overworld changes
- Added mandatory and default connections for inverted in EntranceShuffle.py
- Added general rules, new bunny rules, and new big bomb rules for inverted in Rules.py
    -Changed some dungeon logic for single-entrance dungeons like IP and MM to more closely resemble VT logic, removing rules for doors but compensating with rules for chests
- Link's House is freely shuffled, Dark Sanc is shuffled in West Dark World to strike balance between non-entrance inverted and non-inverted ER
- Old Man spawn point moved to door where his cave exits as in non-entrance inverted, can exit on W DM or W DDM because of portal on W DDM
    -Spawn menu text changes to reflect where this door is
- In shuffles where Aga Tower can be on HC Ledge, Aga Tower is shuffled first or early with one ledge door becoming a must-exit if it's on the ledge to ensure access
- Added pseudo-cave "The Sky" to represent flute connections. Flute works in DW and connects potentially-unreachable regions
- Added exits in LW regions that connect to HC Ledge after Aga is killed

Known Issues:

- Filler occasionally can't place TR keys in some configurations. Most often when Big Key is in the front three chests and a small key is placed in Chain Chomps and Big Chest.
    -Mostly affects dungeonsfull shuffle, comes up rarely in full+
    -Still working on this, can send another patch when it's figured out

### CODE REVIEW

**Comment by cassidoxa** [comment]
The dungeonsfull issue seems to have been solved by placing aga tower off HC Ledge. I think the issue was cropping up when the TR Big Chest door was placed at a must exit entrance on Desert or HC Ledge. Another solution might be to ensure that this exit doesn't connect to Desert Ledge or as the only must-exit on HC Ledge.

**Comment by cassidoxa** [comment]
I think most of the major issues have been ironed out. The TR key issue has been happening for me on the master branch as well so it seems its not unique to inverted. If you're interested in accepting the PR, feel free to let me know if there's any changes you want to see or anything like that and I'd be happy to make them.

**Comment by TheExcalabur** [comment]
Bug: The dark chapel guy still sells you a hint and then does the heal.  This hint prompt should be transferred to sanctuary? (I haven't check what happens without hints yet)
Design query: What's up with the mountain load point? This isn't how either inverted or entrance rando works?  Why aren't we just loading at Old Man House like we usually do?
Design suggestion: Add the Houlihan room and the pyramid crack to the set of rooms to be shuffled.     This could provide cool out-of-logic routing decisions.   Pyramid crack has to be added to the must-exit list.
I haven't yet checked the connection logic in detail to make sure it's bug-free.  This is on my to-do list.

(Good job--glad someone finally put the time in to do this.)

**Comment by cassidoxa** [comment]
The way the mountain spawn is handled in inverted, it puts you at the door where the Old Man Cave exits. For the most part I copied over the inverted stuff in Rom.php from the VT rando so that's how it ended up. This does probably need a better solution. Could keep it at the Old Man House or maybe at Dark Death Mountain Fairy? There's not necessarily a compelling reason to change it from Old Man House in ER I suppose. I'd need to adjust some of the inverted shuffles since I put it in with the regular caves instead of handling it separately as the regular shuffles do.

Moving the Dark Sanc hint to the Sanctuary is probably a good idea. I'm sort of familiar with how the text is handled, but I'm not sure I'd be able to remove healing from Dark Sanc and add a text box to the guy in Sanctuary. I'll look into it but hopefully someone can help out with that bit. And yeah, the pyramid crack is probably better as a must-exit in ER.

Thanks.

**Comment by TheExcalabur** [comment]
I think just standard Old Man House is the right answer for the Old Man spawn--there's no compelling reason to change it.  In non-full shuffles some care might be taken to make sure that it's on DM/DDM.  

Unfortunately I can't help with the text stuff without learning it in the first place.  

Pyramid Crack is actually a mirrorable location in inverted--you can get there from on top of Hyrule castle without having to do anything weird.  

Stuff from code review:
Link and Dark Chapel should be able to connect to must-exits where appropriate—Link’s house to any of them, Dark Chapel none for now.  Can be done just by selecting the entrance from the list of must_exits + entrances, as long as care is taken.

Dark chapel in ‘west dark world’ could include 
“Bumper Cave (Bottom)”
“Bumper Cave (Top)”
“Skull Woods Final Section”
“Thieves Town”
for shuffles that allow multi exit mixing or dungeon mixing.  I think you've already gotten TT where appropriate.  SW back might be a bit weird, but seems fine--if SW is arranged wrong item logic should be able to fix it.  

We should randomly choose if Desert back or Desert left is the must-exit location, as they're logically distinct due to the rocks.  (It'd be nice to choose which side of TR bridge was must-exit too, but it doesn't make a whole ton of difference.)

**Comment by cassidoxa** [comment]
So I've been experimenting with that Pyramid Door. I've figured out how to enter Houlihan from an overworld door, but inverted moves that door to the new HC castle entrance. While I'd like to include it in the shuffle, it looks like it would require some extra rom hacking to add a brand new OW door. I could maybe try it using some empty space but there's the possibility that one of the devs may want to use that space in a future base ROM. And I'm not sure if it's even possible with my current knowledge so I'm going to leave it as it is for now and maybe revisit.

I changed the spawn point back to Old Man Cave in the ROM. So I'll adjust the shuffles to account for this as well. Additionally I'll fix the Dark Sanc and Link's House shuffles and randomize the  Desert must-exit per your suggestions.I'll try to push an update today or tomorrow. 

Thieves' Town is already a possible Dark Sanc location, but it's added in during some shuffles and not in the base variable since it could end up there in simple dungeon shuffles if it was. It's similar to how Tower of Hera is added as an Old Man Entrance in non-inverted. As for Skull Woods, it seems like you can't walk out of there without fire rod-ing the thing in front of the door. Is that how it usually works? My concern with Dark Sanc being there was not having access to West Dark World and Link's House potentially being in a bad spot.

**Comment by TheExcalabur** [comment]
Inverted has one more entrance than the standard rom, which is why they de-linked Houlihan.  I'm surprised something weird needs to be done here, but SNES roms continue to surprise.  Does the crack not two-way connect to Houlihan in the inverted "base" rom?  I'll clone the repo and have a fiddle.  

Do you have a map of the (logical) regions? Trying to reconstruct one from the code is somewhat difficult.

The back of skull woods usually works like other one-way exits--you get kicked out just far enough out that you can move.  Not sure why that would be different.  (Now I want to check that to make sure I'm not telling lies.)  My thought with using SW back as a possible location is that--very occasionally--you wouldn't have early access to WDW; it would also make S&Q slightly less appealing.  It wouldn't happen that often, and I think there's enough resiliency in the system that it wouldn't be too frustrating.

**Comment by cassidoxa** [comment]
The connection from Houlihan is only one-way. I'm not sure about the base ROM but my understanding is that it "moves" the overworld door that was at the pyramid to the one by Hyrule castle. So the game sort of re-uses some data for the former treats the old door like a hard wall. So I think an additional door would require another entry into the exit table specifying the location and camera position and all that. But I'm not sure if there's extra room in the ROM where the current exit table is.

I'll add in the Skull Woods Final Section door as a sanctuary door. I don't have a visual map. I could maybe throw one together in GIMP. I think the main differences are Northeast Light World is its own region due to the bushes and water which the bunny can't cross, IP is just a part of Dark Lake Hylia ( I think I added flippers requirements to the piers so the randomizer doesn't think you can use them if you can exit the IP door without them), DDM is split into Dark Death Mountain and Dark Death Mountain (East Bottom) because of the stairs, and Dark Desert Ledge is its own region since you can only reach the portal with the flute.

Edit: I tested the SW Dark Sanc exit and wasn't able to move out of the door. I guess I'll leave it in for now pending any objections. I'm fine with the 1/11 chance that the player won't have guaranteed access to W DW. I'll have to make sure the logic doesn't assume you can get to West Dark World that way without the fire rod though.

**Comment by TheExcalabur** [comment]
Bug:  Bunny link traversal in the LW isn't quite right.   I rolled a seed (easy-inverted-crystals_crossed-balanced-keysanity_243853716) on the version that was up a few hours ago that put link's house at the overgrown house in Kak.   The logic assumed that I could get from there to the 'Light World' region, which you cannot, in sphere 1.  I'm not sure offhand which other entrances are bush-blocked/liftable-blocked in inverted in such a way that bunny traversal fails, but my guess is that some regions are missing and/or connections don't include the right set of items.  

Bug: the route log seems to also think that I could open the map chest in HC and the chests in paradox having traversed to minimoldorm without a stop in the dark world, so that's another bug.  You can DBR to get the HC chest, but that should be out of logic.  I wonder if something's not quite right with the save point being in the bunny world?

Spoiler log attached

[ER_noglitches_easy-inverted-crystals_crossed-balanced-keysanity_243853716_Spoiler.txt](https://github.com/AmazingAmpharos/ALttPEntranceRandomizer/files/3332078/ER_noglitches_easy-inverted-crystals_crossed-balanced-keysanity_243853716_Spoiler.txt)

**Comment by cassidoxa** [comment]
Thanks. I added a pearl requirement to that house with the bushes, but it may actually need a new region and connections instead.

Edit: For Hyrule Castle and Paradox I think the logic expects you to exit the doors in those locations which lead to the Dark World then re-enter as Link. I've also added a pearl requirement to mini moldorm cave. I thought I'd added it already.

**Comment by cassidoxa** [comment]
Here's maps of the logical regions, with some of the inverted changes notated in blue as well https://imgur.com/a/FKRKFQF

**Comment by cassidoxa** [comment]
To Kevin:

Let me know if there's any changes you'd like to see. I made new keys for Inverted Link's House, Bomb Shop, Aga Tower, and GT in `door_addresses` and `exit_ids` when the vanilla ones could probably be used. The Link's House shuffle was a design choice based on feedback from some cross keys people. That could be changed back to inverted vanilla or refined further. I ended up writing some bytes to the `StartingAreaExit` table for Link's House when I was trying to figure out shuffling that, not sure if it was necessary. Ideally the old Pyramid Entrance could be re-used but that's probably beyond my abilities. I also put the inverted ROM stuff into a new function which only writes some bytes when the shuffle is vanilla. For example the Old Man spawn is on DDM in vanilla but at Old Man House in inverted and the logic accounts for this. 

The flute also uses a new fake cave. This may be buggy, not 100% sure but sometimes it will hang when generating seeds and it seems to be related to the flute. Maybe a recursion issue or something. ~~Lastly, I haven't tested the new `write_int16_bytes` method for the JsonRom object yet.~~

Edit: Also I noticed in the v31 doc that Old Man House overworld doors will be moved to West DDM, so that would need some changes in InvertedRegions.py and maybe the door addresses/exit ids as well.

File: BaseClasses.py
Comment (KevinCathcart) [inline BaseClasses.py]: Perhaps we can add an `is_not_bunny(self, region)` helper, that checks for pearl or light/dark world as applicable. It would simplify checks like this.

File: Dungeons.py
Comment (KevinCathcart) [inline Dungeons.py]: I'm not the biggest fan of all this duplication. If we can come up with a good way to avoid it it would be nice. It is not obvious how best to do that though. See if you can come up with some good way to avoid it. This is not a blocker though.

File: BaseClasses.py
Comment (KevinCathcart) [inline BaseClasses.py]: This could also use that helper, to remove the if.

File: BaseClasses.py
Comment (KevinCathcart) [inline BaseClasses.py]: Let's not add new commented out code.

File: EntranceShuffle.py
Comment (KevinCathcart) [inline EntranceShuffle.py]: I'm torn about the duplicate door_addresses. 
However to the extent that we have them, I'd prefer to see them next to the row they duplicate.

Obviously we cannot completely get rid of them, since Inverted Pyramid Hole has the extra address.

File: EntranceShuffle.py
Comment (KevinCathcart) [inline EntranceShuffle.py]: Technically, the Houlihan exit does not have an entry in the entrance table.

There is an entry for the Houlihan hole target location (which is indeed 0x82), but that is a separate entrance, which really would belong near the bottom on exit_ids. (which is misleadingly named, since it has both entrance and exit ids.

Was there a reason you needed to specify a non-None value here?

File: EntranceShuffle.py
Comment (KevinCathcart) [inline EntranceShuffle.py:3817]: Once again, If we are going to have duplicates, let's keep these duplicates near the entry they duplicate.

File: Rules.py
Comment (KevinCathcart) [inline Rules.py]: Hmm... this is the whirlpool warp from north east dark world to lake hylia. Fake flippers I'm fairly certain fake flippers are possible here (The flippers requirement for glitchless gets added in `no_glitches_rules`.).

However it *is* in the wrong region right now, and if properly put in North East Dark World, this would be simplfied to has_pearl().

File: Rules.py
Comment (KevinCathcart) [inline Rules.py]: Why was this removed in non-inverted? I'm pretty sure we don't want people to be able to mirror from dark world to light world without having a mirror!

File: Rules.py
Comment (KevinCathcart) [inline Rules.py:347]: I'll need to think about these Mire changes.

File: Rom.py
Comment (KevinCathcart) [inline Rom.py]: Umm... what is an int16 byte? Consider renaming as `write_int16s`.

File: Rom.py
Comment (KevinCathcart) [inline Rom.py]: Same.

File: Rom.py
Comment (KevinCathcart) [inline Rom.py]: Surely this could be handled by not setting world.fix_gtower_exit for inverted mode, and thus omitting that check here?

File: Rom.py
Comment (KevinCathcart) [inline Rom.py:765]: I'd prefer to pull out the corresponding inverted writes to these variables, and put them here.

File: Rom.py
Comment (KevinCathcart) [inline Rom.py]: Redundant with line 759

**Review by KevinCathcart — CHANGES_REQUESTED** [review]
I have not reviewed the inverted regions or logic.
But I have looked over most of the rest, and have a few comments/concerns.

File: EntranceShuffle.py
Comment (cassidoxa) [reply EntranceShuffle.py]: In this case I was trying to add the non-inverted Pyramid Entrance back in to the shuffle and needed an extra cave.

File: Rules.py
Comment (cassidoxa) [reply Rules.py:347]: I ported over the VT logic for Mire and Ice Palace, so that's where this came from. My thinking was that as single entrance dungeons it would be a bit simpler to only consider the item locations and the key requirements being moved from doors to chests should prevent unwinnable games.

File: EntranceShuffle.py
Comment (cassidoxa) [reply EntranceShuffle.py]: I made duplicate door_addresses and exit_ids early on to make it a bit easier for me to follow , and I wasn't certain if the door index or room id needed to be changed for inverted or not. If that's not the case then the only reason for duplicates would maybe be for the spoiler logs.

File: Rules.py
Comment (cassidoxa) [reply Rules.py]: Thanks for pointing this out, this was just me being careless with my editor's search.

**Comment by cassidoxa** [comment]
Thanks a lot for the feedback and suggestions. I've tackled most of the low-hanging fruit here. I'll leave it up to you whether the duplicate door_addresses and exit_ids remain. I'll admit that's probably something I should have addressed before now.

File: EntranceShuffle.py
Comment (KevinCathcart) [reply EntranceShuffle.py]: Yeah, since we pyramid exit to the light world, if we were to have something shuffled there, we would need an extra cave. If i were doing that though, I would likely consider adding say an extra instance of the mutliply used fairy fountain or dark world shop. It would be a bit less jarring than entering a cave and finding yourself in the middle of the Houlihan room.

As far as adding that as an extra entrance we would need to add an entry to the overworld doors table to handle it. Technically there is one entry in that that that is unused (it is a duplicate) so it would be possible. But is it worth it for just adding a fairly useless location, since it is not reachable, except by exiting from it?

File: InvertedRegions.py
Comment (KevinCathcart) [inline InvertedRegions.py:166]: Could you also make this change to Regions.py?

**Comment by KevinCathcart** [comment]
Also, please go ahead and mark any resolved comment conversations as resolved. I would do it myself, but i actually don't have push access to AA's repo. Leave any with open questions open obviously.

When I'm ready to pull these changes in, I'll be pulling it into my repo (which is upstream of AA's), but I don't see any reason not to continue the discussion in this pull request thread. 

Generally, I am liking what I am seeing here, and do intend to merge this in at some point. That could be a few more days, and there will probably be a bit more work needed on your part, but I'll get into that a bit later.

File: EntranceShuffle.py
Comment (cassidoxa) [reply EntranceShuffle.py]: The old Pyramid Entrance can be mirrored to from HC ledge in inverted, on the lower left near where it widens out. But yeah, it's not super important to have especially since access to HC ledge is limited as well.

**Comment by cassidoxa** [comment]
Awesome, thanks again for the review.

**Comment by KevinCathcart** [comment]
OK, so having thought things over, I'd actually prefer not to have the dungeon changes. VT's Ice palace logic is actually not completely compatible with monotonicity requirement that ER imposes (this is due to the ternary operator used in that logic).

Other than that I *would* be ready ready to merge this in, except for one thing.

Over  on the Dev branch of my repository, I've pulled in the core set of changes from Bonta0's multiworld fork, in preparation of eventually supporting the official ALTTPR multiworld system.

This makes a lot of changes to support having player numbers all over the place, and it will cause a huge number of merge conflicts with this patch. Also in several spots where no merge conflict would occur, there would still need to be required changes to work properly.

If you could merge in (or rebase on top of) my Dev Branch, and open a pull requerst over in my repository for it, I would greatly appreciate it. (It would likely take me a lot longer to do that myself, given my limited time availability at the moment.)

**Comment by cassidoxa** [comment]
Ok, I'll revert the dungeon changes and work on the new PR. I did try to merge inverted with bonta's multiworld a few weeks ago and only had an issue with the multiworld balancing, which seemed to loop infinitely. So that's the only thing I may need some help with, but I'll ask bonta about it if I can't figure it out before submitting a PR. 

I think what I'll do is pull your dev branch into a new branch of this repo, merge in this branch manually, then submit a PR with that if that's acceptable. I'm still kind of new to git and that seems like the easiest way for me to do it.

**Comment by KevinCathcart** [comment]
Yeah, that would be fine. With the sorts of conflicts that would occur, the automatic merge is not likely to be very helpful, so doing it manually is likely to be easier anyway. 

And don't worry too much about the balancing part. I already know it is somewhat buggy. (It has managed to generate seeds that cannot be beaten, according to ER's own internal Logic checks, at least when only 2 worlds are generated. I've not seen that happen with 3+ worlds). I've not actually looked at that part very closely yet.

If you can get a merged version that works for single world, I'd be more than happy to merge that in pending a determination of why the balancing fails.

**Comment by KevinCathcart** [comment]
Also if you have any questions about what's different in my Dev branch vs Bonta0's or anything like that, please don't hesitate to DM me on Discord. You can find me in the main ALTTPR Discord, and and I'm sure you can figure out my name there. :-D

**Pull request closed without merging**
